### DISCUSSION TITLE
Survey Variables via API

### DISCUSSION THREAD

**Discussion by N3ruru** [semaphoreui/semaphore#1228] [Q&A]
Hello a i need to pass survey Variables via httprequest to semaphore API, i didn't find any example on the API documentation, is it possible to do that? And how? Can you give a body example?

**Comment by slamanna212** [comment]
I am also looking for this and I dont see documentation for it anywhere.

**Comment by ansibleguy** [comment]
If I'm understanding the code correctly, the survey-vars are 'just' written into the `environment` field by the web-frontend.
See: [reference](https://www.ansible-semaphore.com/api-docs/#/project/post_project__project_id__tasks)
Essentially the web-frontend javascript should send the same api-call you need to reproduce.
Would one of you please:
* login to your Semaphore instance
* hit F12 (in most browsers) to open the developer console in your browser
  * click on the 'network' tab
* manually trigger a task using the web-gui
  * check the network log to find the request to `/project/>project_id>/tasks` via POST
  * check the payload sent to the server

Also please share the payload stripped of your personal information!
It could be very helpful to others.

**Comment by N3ruru** [comment]
Thanks a lot, the payload for those who might be interested:
{
    "id": 93,
    "template_id": 7,
    "project_id": 2,
    "status": "success",
    "debug": false,
    "dry_run": false,
    "diff": false,
    "playbook": "",
    "environment": "{\"pwd\":\"\",\"master_pwd\":\"\",\"user\":\"\"}",
    "limit": "",
    "user_id": 5,
    "created": "2023-04-28T10:57:43.075399Z",
    "start": "2023-04-28T10:57:50.338676Z",
    "end": "2023-04-28T10:59:23.04993Z",
    "message": "Criando Storages",
    "commit_hash": null,
    "commit_message": "",
    "build_task_id": null,
    "version": null,
    "arguments": null,
    "tpl_playbook": "playbooks/playbooks_pco_deploy/playbook_pco_deploy.yml",
    "tpl_alias": "[Rebuild Storage] [Company] [Data Center] [QA1]",
    "tpl_type": "",
    "user_name": "<name>",
    "build_task": null
}

**Comment by ansibleguy** [comment]
@N3ruru Thanks.
Looks like `environment` is just a json-string that might get translated into Ansible `extra-vars`
References:
* [runner loading environment](https://github.com/ansible-semaphore/semaphore/blob/develop/services/tasks/runner.go#L397)
* [runner translating environment to extra-vars](https://github.com/ansible-semaphore/semaphore/blob/develop/services/tasks/runner.go#L738) and [building extra-vars from environment](https://github.com/ansible-semaphore/semaphore/blob/develop/services/tasks/runner.go#L621)

 If you get it working using the `environment` field - a working `curl` example would be great! 👍🏼

**Comment by steadfasterX** [comment]
while it is totally true that survey vars will be treated as "normal" environment variables at the end there is a way to also create survey variables over the API!

That becomes very useful as pushing those to the environment will make them "fixed" which removes the flexibility of survey vars.

Here an example CURL call  (assuming you had stored a cookie in `/tmp/semaphore-cookie` before) which will create a template over the API, including 1 survey variable named `debug_off` and nicely formatted override args:

```
curl -v -b /tmp/semaphore-cookie -H 'Content-Type: application/json' -H 'Accept: application/json' http://127.0.0.1:3000/api/project/1/templates -d '{
        "project_id": 1,
        "inventory_id": 1,
        "repository_id": 2,
        "environment_id": 2,
        "vault_key_id": 3,
        "view_id": 4,
        "alias": "my alias",
        "name": "my template name",
        "playbook": "dump.yml",
        "description": "My configuration",
        "arguments": "[\n \"-v\",\n \"-e my_var=myvalue\"\n]",
        "survey_vars": [{"name": "debug_off","title": "Hide debug output? (default: true)","required": false,"type": "","description": "true | false"}],
        "override_args": true
        }'
```

Here a more complex example as an ansible task of the same but 3 survey vars and using an access token. note: the first "set_fact" is not actually needed that way but I find it more readable then specifying it within the body.

```
---

- set_fact:
    zone_template_survey_vars: >-
      [{ 
          "name": "debug_off",
          "title": "Hide debug output? (default: true)",
          "required": false,
          "type": "",
          "description": "true | false"
        },
        {
          "name": "sdteams_skip_notify",
          "title": "Skip any notification? (default false)",
          "required": false,
          "type": "",
          "description": "true or leave empty for default"
        },
        {
          "name": "skip_git_push",
          "title": "Skip pushing changed to git? (default false)",
          "required": true,
          "type": "",
          "description": "true or leave empty for default"
        }]

- name: "Create template(s)"
  ansible.builtin.uri:
    url: "{{ semaphore_api_uri }}/project/{{ semaphore_project_id }}/templates"
    method: POST
    headers:
        Content-Type: application/json
        Authorization: "Bearer {{ vault_semaphore_api_cf_rw }}"
    body_format: json
    body: >-
        { "project_id": 1,
        "inventory_id": 1,
        "repository_id": 2,
        "environment_id": 2,
        "vault_key_id": 3,
        "view_id": 4,
        "alias": "{{ z.name }}",
        "name": "{{ z.name }}",
        "playbook": "dump.yml",
        "description": "My configuration",
        "arguments": "[\n \"-v\",\n \"-e myvar={{ z.name }}\"\n]",
        "survey_vars": {{ zone_template_survey_vars }},
        "override_args": true }
    status_code: [ 201, 200 ]
  register: tmplc
  changed_when: tmplc.status == 201
  loop: "{{ cf_all_zones.json.result }}"
  loop_control:
    loop_var: z
  when: z.name not in sp_all_templates.json | map(attribute='name') | list
```

Of course that one is a more complex one as I iterate over a list (and filter out only those I need) but I guess you get the idea.

**Comment by ansibleguy** [comment]
@steadfasterX Thank you for the example for **creating templates** via API.

I think @N3ruru asked how to pass survey_vars when **starting tasks** via API. In that case there is no `survey_vars` json-field.
The `environment` json-field for **starting tasks** is not 'static' - its content can differ between runs. (_not the 'environments' found in the WebUI sub-option - it's just the name of the JSON field_)

**Reply by steadfasterX** [reply]
actually it is not clear to me if he wants to **start** or **create**. anyways those ppl (like me) coming here bc of the title of this discussion might find it useful

**Reply by ansibleguy** [reply]
I agree. 👍🏼
Just wanted to clarify
